### PR TITLE
Make the movement of the Grid columns really easy

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | While it's already possible, swapping 2 existing columns is boring and error-prone with the current API. Let's create a specific function for that as the need is really common.
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | yes, some developers will lose their job as it's now super easy to do that /c @MDWeb-Lille 
| Fixed ticket? | none
| How to test?  | There is nothing to do from a QA point of view, but we MUST update every docs and demonstration modules and training resources, starting from 1.7.7!

### Before

```php
// for a grid with 2 columns "a", "b" and "c".
$columnCollection
    ->remove('a')
    ->addAfter(
        'b',
        (new DataColumn('a'))
            ->setName('A')
            ->setOptions([...])
   );
// In this case, we must create again column A for no reasons!
// And we need to do it X times, X is the number of swaps
```

### After

```php
// Remember, the collection starts at position 0
$columnCollection->move('a', 1) // Column 'a' is now in position 2
    ->move('b', 2) // calls can be chained
    ->move('c', 0)
;
// Now, go grab a coffee or play baby foot
```

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/13803)

### CODE REVIEW

**Review by matks — DISMISSED** [review]

File: src/Core/Grid/Column/ColumnCollection.php [hunk 1]
```diff
@@ -107,14 +107,39 @@ public function toArray()
...
 
+    /**
+     * Swap two existing columns.
+     *
+     * @param string $keyFrom the Column ID original position in the Collection
+     * @param string $keyTo the Column ID destination position in the Collection
+     *
+     * @return self
+     */
+    public function swap($keyFrom, $keyTo)
```
Comment (jolelievre) [inline src/Core/Grid/Column/ColumnCollection.php]: So you don't change the interface? Isn't it a BC break to add a public method in this class?

File: src/Core/Grid/Column/ColumnCollection.php [hunk 2]
```diff
@@ -107,14 +107,39 @@ public function toArray()
...
+    {
+        if (!array_key_exists($keyFrom, $this->items) || !array_key_exists($keyTo, $this->items)) {
+            throw new ColumnNotFoundException(sprintf(
+                'Cannot swap two existing columns from collection. Columns with ids "%s" or "%s" were not found.',
```
Comment (jolelievre) [inline src/Core/Grid/Column/ColumnCollection.php]: The error message is confusing, you can't say "two existing columns" when at least one is missing

File: src/Core/Grid/Column/ColumnCollection.php (hunk 1 above)
Comment (mickaelandrieu) [reply src/Core/Grid/Column/ColumnCollection.php]: why? changing the interface is a BC but a class doesn't have to follow strictly its interface.

Of course, we shouldn't use `swap` function in the core as we have type hinted everywhere with the interface: but why would we? :D This function is for the developers of modules and the object they receive from us is always a [ColumnCollection](https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Definition/Factory/CategoryGridDefinitionFactory.php#L122) instance.

File: src/Core/Grid/Column/ColumnCollection.php (hunk 2 above)
Comment (mickaelandrieu) [reply src/Core/Grid/Column/ColumnCollection.php]: This is what happens when we try to build exception messages in classes not built for :/ 

Honest: I think it's good enough for a developer to figure out which id has been malformed but I'm open to suggestions **except** throwing 2 exceptions or making the building of the message bigger in a class not in charge of that.

What do you suggest?

**Comment by mickaelandrieu** [comment]
As a note: @PrettyCI is down...

File: src/Core/Grid/Column/ColumnCollection.php [hunk 3]
```diff
@@ -107,14 +107,39 @@ public function toArray()
...
+     * Swap two existing columns.
+     *
+     * @param string $keyFrom the Column ID original position in the Collection
+     * @param string $keyTo the Column ID destination position in the Collection
+     *
+     * @return self
+     */
+    public function swap($keyFrom, $keyTo)
+    {
+        if (!array_key_exists($keyFrom, $this->items) || !array_key_exists($keyTo, $this->items)) {
```
Comment (sarjon) [inline src/Core/Grid/Column/ColumnCollection.php]: Wouldn't it be more performant to check `!isset($this->items[$keyFrom], $this->items[$keyTo])`?

File: src/Core/Grid/Column/ColumnCollection.php (hunk 1 above)
Comment (sarjon) [reply src/Core/Grid/Column/ColumnCollection.php]: I'd prefer not to use `key`, but `id` instead to keep consistency between other methods.

Comment (jolelievre) [reply src/Core/Grid/Column/ColumnCollection.php]: Fine for the "BC break"
Indeed using id would be more consistent with all the other functions parameters

File: src/Core/Grid/Column/ColumnCollection.php (hunk 2 above)
Comment (jolelievre) [reply src/Core/Grid/Column/ColumnCollection.php]: I'd replace `Cannot swap two existing columns from collection` with `Cannot swap nonexistent columns from collection` which seems clearer

File: src/Core/Grid/Column/ColumnCollection.php (hunk 1 above)
Comment (jolelievre) [reply src/Core/Grid/Column/ColumnCollection.php]: You even use "id" in your exception message

Comment (mickaelandrieu) [reply src/Core/Grid/Column/ColumnCollection.php]: yup, fair enough: let me update

File: src/Core/Grid/Column/ColumnCollection.php (hunk 3 above)
Comment (mickaelandrieu) [reply src/Core/Grid/Column/ColumnCollection.php]: I did it first, and my tests were broken. I don't know why but I believe in unit tests more than in `isset()` function ;)

File: src/Core/Grid/Column/ColumnCollection.php (hunk 2 above)
Comment (mickaelandrieu) [reply src/Core/Grid/Column/ColumnCollection.php]: I've updated: wdyt?

Comment (mickaelandrieu) [reply src/Core/Grid/Column/ColumnCollection.php]: Solved as we have changed from swap() to move()

File: src/Core/Grid/Column/ColumnCollection.php [hunk 4]
```diff
@@ -107,14 +107,43 @@ public function toArray()
...
+            throw new ColumnNotFoundException(sprintf(
+                'Cannot insert new column into collection. Column with id "%s" was not found.',
+                $id
+            ));
+        }
+
+        $column = $this->items[$id];
+        unset($this->items[$id]);
+
+        $columns = array_slice($this->items, 0, $position - 1, true) +
```
Comment (jolelievre) [inline src/Core/Grid/Column/ColumnCollection.php]: I don't think index position on 1 is a good idea. It is confusing for developers

File: src/Core/Grid/Column/ColumnCollection.php [hunk 5]
```diff
@@ -107,14 +107,43 @@ public function toArray()
...
+            [$column->getId() => $column] +
+            array_slice($this->items, $position - 1, $this->count(), true);
```
Comment (jolelievre) [inline src/Core/Grid/Column/ColumnCollection.php]: Shouldn't length be `$this->count() - $position` or something like that?

File: tests-legacy/Unit/Core/Grid/Column/ColumnCollectionTest.php [hunk 6]
```diff
@@ -215,4 +260,15 @@ private function getColumnPositions(ColumnCollection $columns)
...
         return $positions;
     }
+
+    /**
+     * Helper assertion.
+     *
+     * @param ColumnCollection $columnCollection
+     * @param string $columnId
+     */
+    private function isColumnWithId(ColumnCollection $columnCollection, $columnId)
```
Comment (jolelievre) [inline tests-legacy/Unit/Core/Grid/Column/ColumnCollectionTest.php]: I'd name this `assertColumnId` wdyt?

**Review by jolelievre — DISMISSED** [review]

File: src/Core/Grid/Column/ColumnCollection.php (hunk 4 above)
Comment (mickaelandrieu) [reply src/Core/Grid/Column/ColumnCollection.php]: its fixed

File: src/Core/Grid/Column/ColumnCollection.php (hunk 5 above)
Comment (mickaelandrieu) [reply src/Core/Grid/Column/ColumnCollection.php]: fixed

File: tests-legacy/Unit/Core/Grid/Column/ColumnCollectionTest.php (hunk 6 above)
Comment (mickaelandrieu) [reply tests-legacy/Unit/Core/Grid/Column/ColumnCollectionTest.php]: make sense, I'll update that

**Review by jolelievre — APPROVED** [review]
Nice job @mickaelandrieu

**Comment by jolelievre** [comment]
Thanks @mickaelandrieu

**Pull request merged**